src/event-stream/tokens-contract-handler.ts

```diff
@@ -0,0 +1,417 @@
+   */
+  private async handleFTContract() {
+    try {
+      //make read-only call to contract
+      const contractCallName = await this.makeReadOnlyContractCall('get-name', []);
+      const contractCallUri = await this.makeReadOnlyContractCall('get-token-uri', []);
+      const contractCallSymbol = await this.makeReadOnlyContractCall('get-symbol', []);
+      const contractCallDecimals = await this.makeReadOnlyContractCall('get-decimals', []);
+
+      //check parse clarity values
+      const nameCV = this.checkAndParseString(contractCallName);
+      const uriCV = this.checkAndParseOptionalString(contractCallUri);
+      const symbolCV = this.checkAndParseString(contractCallSymbol);
+      const decimalsCV = this.checkAndParseUintCV(contractCallDecimals);
+
+      let metadata: FTTokenMetadata = { name: '', description: '', image: '' };
+      //fetch metadata from the uri if possible
+      if (uriCV) metadata = await this.getMetadataFromUri<FTTokenMetadata>(uriCV.data);
+      const { name, description, image } = metadata;
+      const fungibleTokenMetadata: DbFungibleTokenMetadata = {
+        name: nameCV ? nameCV.data : name, //prefer the on-chain name
+        description: description,
+        image_uri: image,
+        image_canonical_uri: uriCV ? uriCV.data : '',
+        symbol: symbolCV ? symbolCV.data : '',
+        decimals: decimalsCV ? Number(decimalsCV.value) : 0,
+        contract_id: `${this.contractAddress}.${this.contractName}`,
+      };
+
+      //store metadata in db
+      await this.storeFTMetadata(fungibleTokenMetadata);
+    } catch (error) {
+      logger.error('erro: error handling FT contract', error);
+    }
+  }
+
+  /**
+   * fetch Non Fungible contract metadata
+   */
+  private async handleNFTContract() {
+    try {
+      const contractCallTokenId = await this.makeReadOnlyContractCall('get-last-token-id', []);
+      const tokenId = this.checkAndParseUintCV(contractCallTokenId);
+
+      if (tokenId) {
+        const contractCallUri = await this.makeReadOnlyContractCall('get-token-uri', [tokenId]);
+        const uriCV = this.checkAndParseOptionalString(contractCallUri);
+
+        let metadata: NFTTokenMetadata = { name: '', description: '', imageUrl: '' };
+        if (uriCV) metadata = await this.getMetadataFromUri<NFTTokenMetadata>(uriCV.data);
+
+        const nonFungibleTokenMetadata: DbNonFungibleTokenMetadata = {
+          name: metadata.name,
+          description: metadata.description,
+          image_uri: metadata.imageUrl,
+          image_canonical_uri: uriCV ? uriCV.data : '',
+          contract_id: `${this.contractAddress}.${this.contractName}`,
+        };
+
+        await this.storeNFTMetadata(nonFungibleTokenMetadata);
+      }
+    } catch (error) {
+      logger.error('erro: error handling NFT contract', error);
+    }
+  }
+
+  /**helpng method for creating http url */
+  private makeHostedUrl(uri: string): string {
+    if (uri.includes('http')) return uri;
```

Comment:
We should try to use the native `URL` API rather than another lib `uri-js`. And we should explicitly check for the supported protocols (http, https, ipfs). 
Then this code would look something like 
```ts
const parsedUri = new URL(uri);
if (parsedUri.protocol === 'http:' || parsedUri.protocol === 'https:') return uri;
if (parsedUri.protocol === 'ipfs:') { .... }
throw new Error(`Unsupported uri protocol: ${uri}`);
```

Reply:
isn't NFT immutable?